code/drasil-lang/lib/Language/Drasil/Expr/Lang.hs

```diff
@@ -317,178 +333,183 @@ instance Typed Expr Space where
-      then Left lt
-      else Right $ "Set contains should only be applied to Set of same space. Received `" ++ show lt ++ "` / `" ++ show rt ++ "`."
-    (_      , Right e) -> Right e
-    (Right e, _      ) -> Right e
-    (Left lt, Left rsp) -> Right $ "Set union expects set operands. Received `" ++ show lt ++ "` · `" ++ show rsp ++ "`."
-
-  infer cxt (Operator _ (S.BoundedDD _ _ bot top) body) =
-    let expTy = S.Integer in
-    case (infer cxt bot, infer cxt top, infer cxt body) of
-      (Left botTy, Left topTy, Left bodyTy) -> if botTy == S.Integer
-        then if topTy == S.Integer
-          then if expTy == bodyTy
-            then Left expTy
-            else Right $ "'Big' operator range body not Integer, found: " ++ show bodyTy ++ "."
-          else Right $ "'Big' operator range top not Integer, found: " ++ show topTy ++ "."
-        else Right $ "'Big' operator range bottom not of expected type: " ++ show expTy ++ ", found: " ++ show botTy ++ "."
-      (_         , _         , Right x    ) -> Right x
-      (_         , Right x   , _          ) -> Right x
-      (Right x   , _         , _          ) -> Right x
-
-  infer cxt (RealI uid ri) =
-    case (inferFromContext cxt uid, riTy ri) of
-      (Left S.Real, Left riSp) -> if riSp == S.Real
-        then Left S.Boolean
-        else Right $
-          "Real interval expects interval bounds to be of type Real, but received: " ++ show riSp ++ "."
-      (Left uidSp, _         ) -> Right $
-        "Real interval expects variable to be of type Real, but received `" ++ show uid ++ "` of type `" ++ show uidSp ++ "`."
-      (_          , Right x  ) -> Right x
-      (Right x    , _        ) -> Right x
+
+  infer cxt (VVNBinaryOp Dot l r) = do
+    lt <- infer cxt l
+    rt <- infer cxt r
+    let msg hand sp = "Vector dot product expects a numeric vector on the " ++ hand ++ "-hand side, but found `" ++ sp ++ "`."
+    lvet <- assertNumericVector lt (msg "left")
+    rvet <- assertNumericVector rt (msg "right")
+    assertEquivNumeric lvet rvet
+      (\ls rs -> "Vector dot product expects vectors of the same numeric type, but found `" ++ ls ++ "` and `" ++ rs ++ "`.")
+    pure lvet
+
+  infer cxt (ESSBinaryOp _ l r) = do
+    lt <- infer cxt l
+    rt <- infer cxt r
+    set <- assertSet rt
+      (\sp -> "Set add/subtract expects a set on the right-hand side, but found `" ++ sp ++ "`.")
+    assertEquivNumeric lt set
+      (\ls rs -> "Set add/subtract expects numeric set operands. Received `" ++ ls ++ "` / `" ++ rs ++ "`.")
+    pure rt
+
+  infer cxt (ESBBinaryOp SContains l r) = do
+    lt <- infer cxt l
+    rt <- infer cxt r
+    set <- assertSet rt
+      (\sp -> "Set contains expects a set on the right-hand side, but found `" ++ sp ++ "`.")
+    assertEquivNumeric lt set
+      (\ls rs -> "Set contains should only be applied to Set of numeric type. Received `" ++ ls ++ "` / `" ++ rs ++ "`.")
+    pure S.Boolean
```

Comment:
I don't actually see why for `Set`s in particular, its reasonable to assume they're always sets of numbers, but I followed logic. A subsequent PR can alter the rule.